### PR TITLE
[ShapeOpt] tests don't write output by default

### PR DESCRIPTION
**Pull request by armingeiser**
It was suggested in one of the previous MRs:

The tests do not delete the results after a succesful run, instead the result files are simply added to `.gitignore`. This makes it more convenient to actually have a look at the test results, e.g. for debugging or similar.

### CODE REVIEW

**Review by philbucher — CHANGES_REQUESTED** [review]
tests should by default NOT create post-output!
consider that 99.9% the tests run is in the CI, where output that is not used for checking the results is wasted time => time that we are usually short of!

usually the tests have an option with which the output can be enabled
e.g. like [this](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/tests/test_laplacian_mesh_motion_2d.py#L24-L26)

I recommend doing sth in this direction, how exactly is up to you

**Comment by armingeiser** [comment]
The tests are using the output and compare the values from the resulting files - `optimization_history.csv`. So we will need that one.

I agree however that the post-output is not used, i will see how i can omit it.

**Comment by philbucher** [comment]
ok for me now
I will release my block once @dbaumgaertner approved

File: applications/ShapeOptimizationApplication/tests/.gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+**/Optimization_Results/
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/.gitignore:1]: Is there a reason why you deleted the "delete"-command from the test runs and added the output to the gitignore? I personally find it much cleaner to delete all unnecessary output instead of ignoring it.

File: applications/ShapeOptimizationApplication/tests/.gitignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+**/response_combination.csv
+
+**/Optimization_Results*/
+**/algorithm_gradient_projection_test
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/.gitignore]: There is no such test case. The same holds for the following linen.

File: applications/ShapeOptimizationApplication/tests/algorithm_bead_optimization_test/run_test.py [hunk 3]
```diff
@@ -51,10 +51,3 @@
     TestCase().assertAlmostEqual(resulting_penalty_factor, 3.26430E+01,5)
 
 os.chdir(original_directory)
-
-# Cleaning
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/algorithm_bead_optimization_test/run_test.py:55]: See my comment above. Why did you remove all the lines? Without writing the design surface we still need the first two lines, right?

There are more files following where this comment applies.

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json [hunk 4]
```diff
@@ -63,10 +63,19 @@
...
+            "nodal_results"      : [
+                "SHAPE_CHANGE",
+                "DF1DX",
+                "DF1DX_MAPPED",
+                "DC1DX",
+                "DC1DX_MAPPED",
+                "SEARCH_DIRECTION",
+                "SHAPE_UPDATE"
+            ],
             "output_format" : {
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json:67]: It would make sense to not require the output_format settings if `write_design_outout=false`.

File: applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py [hunk 5]
```diff
@@ -46,6 +46,7 @@ def __init__( self, ModelPartController, Communicator, OptimizationSettings ):
             "optimization_log_filename" : "optimization_log",
             "design_output_mode"        : "WriteOptimizationModelPart",
             "nodal_results"             : [ "SHAPE_CHANGE" ],
+            "write_design_output"       : true,
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]: I would say there is already an option that could be used for this purpose. Just see the option "design_output_mode" above. You could just introduce another case for this option which is called something like "NoDesignOutput". What do you think?

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json [hunk 6]
```diff
@@ -63,10 +63,19 @@
         },
         "output" : {
             "design_output_mode" : "WriteDesignSurface",
-            "nodal_results"      : [ "SHAPE_CHANGE" ],
+            "nodal_results"      : [
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: Is there a reason why you added more output? I thought the less the better in case of the tests.

File: applications/ShapeOptimizationApplication/tests/algorithm_trust_region_test/parameters.json [hunk 7]
```diff
@@ -64,10 +64,11 @@
         },
         "output" : {
             "design_output_mode" : "WriteDesignSurface",
-            "nodal_results"      : [ "SHAPE_CHANGE" ],
+            "nodal_results"      : [ "SHAPE_CHANGE", "SHAPE_UPDATE" ],
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/algorithm_trust_region_test/parameters.json]: Again, why more output?

**Review by dbaumgaertner — CHANGES_REQUESTED** [review]
Thanks for taking care of it. I have some questions/suggestions here and there.

File: applications/ShapeOptimizationApplication/tests/.gitignore (hunk 1 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/.gitignore:1]: I often use tests in order to compare new developments with the "tested" stuff. The fact that the results are deleted after a successful run is annoying. IMO deleting the results after the test is doing extra stuff that is not part of the default behavior of the code. But this may be my personal preference.

File: applications/ShapeOptimizationApplication/tests/.gitignore (hunk 2 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/.gitignore]: accidentally commited

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 4 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json:67]: The idea here is that it should be easy to switch on the output for each test whenever needed.
By removing the related part one has to copy the settings (that typically no one knows by heart from somewhere) thats a lot of extra effort.

File: applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py (hunk 5 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]: Good Idea :+1:

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 6 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: The less the better is now no output (by default). That would allow us to have the full output that makes sense for this case prepared and then simply switch it on when needed.

File: applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py (hunk 5 above)
Comment (philbucher) [reply applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]: obviously you mean `no_design_output` right :smile:

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]: Having a quick look at our json files, i think this is the last remaining setting in ShapeOpt that uses CamelCase naming. I could change it if you agree @dbaumgaertner

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 6 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: however that is a different discussion and i revert it here

File: applications/ShapeOptimizationApplication/tests/algorithm_trust_region_test/parameters.json (hunk 7 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/algorithm_trust_region_test/parameters.json]: reverted.

**Comment by armingeiser** [comment]
@dbaumgaertner do you have any more comments?

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json [hunk 8]
```diff
@@ -63,19 +63,10 @@
         },
         "output" : {
             "design_output_mode" : "WriteDesignSurface",
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: Shouldn't now in all unit test "design_output_mode" be "None"? Or am I missing something? I mean that was the original reason to introduce this option. If yes, then it is missing in all algorithm tests.

File: applications/ShapeOptimizationApplication/tests/in_plane_opt_test/optimization_parameters.json [hunk 9]
```diff
@@ -57,12 +57,11 @@
             }
         },
         "output" : {
-            "design_output_mode" : "WriteDesignSurface",
+            "design_output_mode" : "None",
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/in_plane_opt_test/optimization_parameters.json]: Here it is "None"

File: applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py [hunk 10]
```diff
@@ -74,11 +73,18 @@ def __CreateValueLogger( self ):
 
     # -----------------------------------------------------------------------------
     def __CreateDesignLogger( self ):
-        outputFormatName = self.OptimizationSettings["output"]["output_format"]["name"].GetString()
-        if not self.OptimizationSettings["output"]["write_design_output"].GetBool():
+        valid_output_modes = ["WriteDesignSurface", "WriteOptimizationModelPart", "None"]
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]: Much cleaner now.

**Review by dbaumgaertner — COMMENTED** [review]
Just one more thing, see my comment. Other than that I would approve it.

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 8 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: I left one gid and one vtk output active on purpose, such that they are tested as well

**Review by dbaumgaertner — DISMISSED** [review]

File: applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py [hunk 11]
```diff
@@ -73,6 +73,17 @@ def __CreateValueLogger( self ):
 
     # -----------------------------------------------------------------------------
     def __CreateDesignLogger( self ):
+        valid_output_modes = ["WriteDesignSurface", "WriteOptimizationModelPart", "None"]
```
Comment (philbucher) [inline applications/ShapeOptimizationApplication/python_scripts/data_logger_factory.py]:
```suggestion
        valid_output_modes = ["WriteDesignSurface", "WriteOptimizationModelPart", "write_design_surface","write_optimization_model_part", "none"]
```

this way you can make it backward compatible

**Review by philbucher — DISMISSED** [review]

File: applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json (hunk 8 above)
Comment (philbucher) [reply applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/parameters.json]: You could add a test if it is written correctly (maybe also add it to the tests of the `DesingLogger`)
otherwise all this does it check that there are no exceptions

there are [functionalities in the core](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/compare_two_files_check_process.py#L88-L104) to compare output files

**Review by philbucher — APPROVED** [review]
:heart:

**Comment by dbaumgaertner** [comment]
@armingeiser now you switched off all the tests. So the output is not tested. My thumbs up above meant that I forgot about this aspect, and you were right, we should leave one test with gid and one with vtk output. What do you think?

**Comment by armingeiser** [comment]
I will create a test for the output specifically as @philbucher suggested.

**Comment by loumalouomega** [comment]
@armingeiser can we delete the branch?
